cluster-autoscaler/core/scaledown/planner/planner.go

```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		p.CleanUpUnneededNodes()
+		return errors.ToAutoscalerError(errors.UnexpectedClusterStateError, err)
+	}
+	deletions := asMap(merged(as.DeletionsInProgress()))
+	podDestinations = filterOutOngoingDeletions(podDestinations, deletions)
+	scaleDownCandidates = filterOutOngoingDeletions(scaleDownCandidates, deletions)
+	p.categorizeNodes(asMap(nodeNames(podDestinations)), scaleDownCandidates, pdb)
+	p.rs.DropOldHints()
+	p.scheduler.DropOldHints()
+	return nil
+}
+
+// CleanUpUnneededNodes forces Planner to forget about all nodes considered
+// unneeded so far.
+func (p *Planner) CleanUpUnneededNodes() {
+	p.unneededNodes.Clear()
+}
+
+// NodesToDelete returns all Nodes that could be removed right now, according
+// to the Planner.
+func (p *Planner) NodesToDelete() (empty, needDrain []*apiv1.Node, result status.ScaleDownResult, aErr errors.AutoscalerError) {
+	nodes, err := allNodes(p.context.ClusterSnapshot)
+	if err != nil {
+		return nil, nil, status.ScaleDownError, errors.ToAutoscalerError(errors.InternalError, err)
+	}
+	resourceLimiter, err := p.context.CloudProvider.GetResourceLimiter()
+	if err != nil {
+		return nil, nil, status.ScaleDownError, errors.ToAutoscalerError(errors.CloudProviderError, err)
+	}
+	limitsLeft := p.resourceLimitsFinder.LimitsLeft(p.context, nodes, resourceLimiter, p.latestUpdate)
+	empty, needDrain, unremovable := p.unneededNodes.RemovableAt(p.context, p.latestUpdate, limitsLeft, resourceLimiter.GetResources(), p.actuationStatus)
+	for _, u := range unremovable {
+		p.unremovableNodes.Add(u)
+	}
+	// TODO: filter results with ScaleDownSetProcessor.GetNodesToRemove
+	return empty, needDrain, status.ScaleDownNodeDeleteStarted, nil
+}
+
+func allNodes(s clustersnapshot.ClusterSnapshot) ([]*apiv1.Node, error) {
+	nodeInfos, err := s.NodeInfos().List()
+	if err != nil {
+		// This should never happen, List() returns err only because scheduler interface requires it.
+		return nil, err
+	}
+	nodes := make([]*apiv1.Node, len(nodeInfos))
+	for i, ni := range nodeInfos {
+		nodes[i] = ni.Node()
+	}
+	return nodes, nil
+}
+
+// UnneededNodes returns a list of nodes currently considered as unneeded.
+func (p *Planner) UnneededNodes() []*apiv1.Node {
+	return p.unneededNodes.AsList()
+}
+
+// UnremovableNodes returns a list of nodes currently considered as unremovable.
+func (p *Planner) UnremovableNodes() []*simulator.UnremovableNode {
+	return p.unremovableNodes.AsList()
+}
+
+// NodeUtilizationMap returns a map with utilization of nodes.
+func (p *Planner) NodeUtilizationMap() map[string]utilization.Info {
+	return p.nodeUtilizationMap
+}
+
+func (p *Planner) verifyOngoingActuation() error {
+	err := p.verifyPodsCanSchedule(currentlyDrainedPods(p.context.ClusterSnapshot.NodeInfos(), p.actuationStatus))
+	if err != nil {
+		return err
+	}
+	// TODO(x13n): Check owner references to avoid double-counting already
+	// recreated pods.
+	return p.verifyPodsCanSchedule(p.actuationStatus.RecentEvictions())
```

Comment:
Got it, definitely not arguing with being on the safe side here. Just from the perspective of reading the code - a comment explaining that the second check may be redundant but that's intended for safety/stability would go a long way (otherwise my first instinct is to find out the difference between the two sets of pods and why they're possibly intersecting). Leaving this as a nit.

Reply:
Explaining why the code is doing what it is doing sounds good to me - added.